File: 10-逆向工程/04-固件与嵌入式逆向/JFFS2与UBI闪存结构/go/main.go
package main

import (
	"fmt"

	"jffs2ubi"
)

func main() {
	n := &jffs2ubi.RawNode{Magic: jffs2ubi.MagicBitmask,
		Nodetype: jffs2ubi.NodetypeInode, Totlen: 100}
	jffs2ubi.SealCommon(n)
	raw := n.Common()
	p, _ := jffs2ubi.ParseCommon(raw, 0)
	fmt.Printf("jffs2 magic=%#06x type=%#06x totlen=%d hdr_crc_ok=%v incompat=%v\n",
		p.Magic, p.Nodetype, p.Totlen, p.CheckHdrCRC(), jffs2ubi.IsIncompat(p.Nodetype))

	ec := jffs2ubi.PackEC(&jffs2ubi.ECHeader{
		Magic: jffs2ubi.UBIEcHdrMagic, Version: jffs2ubi.UBIVersion,
		Ec: 7, VidHdrOffset: 2048, DataOffset: 4096, ImageSeq: 0x11223344})
	fmt.Printf("ubi ec magic=%#010x size=%d crc_ok=%v\n",
		uint32(ec[0])<<24|uint32(ec[1])<<16|uint32(ec[2])<<8|uint32(ec[3]),
		len(ec), jffs2ubi.CheckECCRC(ec))

	vid := jffs2ubi.PackVID(&jffs2ubi.VIDHeader{
		Magic: jffs2ubi.UBIVidHdrMagic, Version: jffs2ubi.UBIVersion,
		VolType: jffs2ubi.UBIVidDynamic, VolID: 0, Lnum: 3, Sqnum: 11})
	fmt.Printf("ubi vid size=%d crc_ok=%v layout_id=%#x\n",
		len(vid), jffs2ubi.CheckVIDCRC(vid), jffs2ubi.LayoutVolumeID)

	old := &jffs2ubi.VIDHeader{Sqnum: 10}
	nw := &jffs2ubi.VIDHeader{Sqnum: 20, CopyFlag: 1}
	fmt.Printf("copy+crc_bad -> old=%v  copy+crc_ok -> new=%v\n",
		jffs2ubi.PickPEB(old, nw, false) == old,
		jffs2ubi.PickPEB(old, nw, true) == nw)
}

File: 10-逆向工程/04-固件与嵌入式逆向/JFFS2与UBI闪存结构/python/main.py
"""demo593 演示入口：扫一串 JFFS2 结点，再看 UBI 的 EC/VID 头。"""

import jffs2ubi as ju


def main():
    data = b"\xde\xad\xbe\xef"
    stream = b""
    stream += ju.build_dirent(1, 2, 42, 1700000000, 3, 4, b"etc")
    stream += ju.build_raw_inode(42, 5, 0o100644, 0, 0, len(data), 0,
                                 len(data), len(data), ju.JFFS2_COMPR_ZLIB, data)

    print("== JFFS2 结点流 ==")
    off = 0
    while off + ju.UNKNOWN_NODE_SIZE <= len(stream):
        n = ju.parse_common(stream, off)
        kind = "INODE" if n.nodetype == ju.JFFS2_NODETYPE_INODE else \
            ("DIRENT" if n.nodetype == ju.JFFS2_NODETYPE_DIRENT else "?")
        print("  off=%-4d type=%#06x %-6s totlen=%-4d hdr_crc=%s"
              % (off, n.nodetype, kind, n.totlen, ju.check_hdr_crc(n)))
        off += n.totlen

    print("== UBI EC/VID ==")
    ec = ju.pack_ec(ju.EcHeader(ec=7, vid_hdr_offset=2048, data_offset=4096,
                                image_seq=0x11223344))
    e = ju.parse_ec(ec)
    print("  EC magic=%#x ec=%d vid_off=%d data_off=%d crc_ok=%s"
          % (e.magic, e.ec, e.vid_hdr_offset, e.data_offset, ju.check_ec_crc(ec)))
    vid = ju.pack_vid(ju.VidHeader(vol_id=0, lnum=3, sqnum=11,
                                   vol_type=ju.UBI_VID_DYNAMIC))
    v = ju.parse_vid(vid)
    print("  VID magic=%#x vol_id=%d lnum=%d sqnum=%d crc_ok=%s"
          % (v.magic, v.vol_id, v.lnum, v.sqnum, ju.check_vid_crc(vid)))
    print("  layout volume id = %#x" % ju.UBI_LAYOUT_VOLUME_ID)


if __name__ == "__main__":
    main()
